### PR TITLE
Fix : issue#4596, a TypeError, and a Node not found error from 'getNode'

### PR DESCRIPTION
**Pull request by zqqcee**
## Checklist

- [ x ] `npm test` passes
- [ x ] commit message follows commit guidelines

### To fix the bug #4596 , I've changed packages/g6/src/item/item.ts
**old version:**
```typescript
const {
     type = this.type === 'node' ? 'circle-node' : 'line-edge',
     lodStrategy: modelLodStrategy,
   } = this.displayModel.data;
```

**new version:**
```typescript
const type = this.type === 'node' ? 'circle-node' : 'line-edge';
const {
      lodStrategy: modelLodStrategy,
    } = this.displayModel.data;
```

### To fix the bug Node not found, I've changed packages/g6/src/runtime/controller/item.ts
image

**old version:**
```typescript
const innerModel = graphCore.getNode(id);
const relatedEdgeInnerModels = graphCore.getRelatedEdges(id);
```

**new version:**
```typescript
// 'id' variable is always string in here, but one in user data is number, possibly.        
let innerModel
try {
  innerModel = graphCore.getNode(id);
} catch (e) {
  innerModel = graphCore.getNode(Number(id))
}

let relatedEdgeInnerModels;
try {
  relatedEdgeInnerModels = graphCore.getRelatedEdges(id);
} catch (error) {
  relatedEdgeInnerModels = graphCore.getRelatedEdges(Number(id));
}
```

---
### For the bug #4596
When the data imported by the user has a "type" attribute, like this:
```typescript
const data = {
        nodes: [
            {
                id: 1,
                data: {
                    name: 'node1',
                    x: 100,
                    y: 100,
                    type: 'circle'
                }
            },
            {
                id: 2,
                data: {
                    name: 'node2',
                    x: 200,
                    y: 100,
                    type: 'circle'
                }
            },
        ],
    };
```
Putting "type" in "{}" for object destructuring will cause the following problems:
the `type` variable in `packages/g6/src/item/item.ts ,line117` **will be overwritten!**
It will cause `type` failing to be the expected type: 'circle-node' or 'line-edge', which further causes `RenderExtension` to be undefined.

### For the bug Node not found mentioned before
An error is throwed from this function 'getNode()'; msg: "Node not found for id: 1"
After testing, I found that 'id' variable is always string after `Object.keys(update).forEach((id)=>{...})`, but one in user data is number, possibly.    
I tried adding a `try-catch` block to fix this bug, but **it doesn't seem very reasonable to do so**. I guess you can restrict the user to set the `id` to a string type in the document, or force the `id` to a string type in the `transformer data` layer to avoid this error.

This PR, 
- moving the definition of the type variable out of the parentheses solves the bug #4596 .
- adding `try-catch` to solves the bug : "Node not found for id" 
The test demo located in here:  packages/g6/tests/intergration/bugReproduce.ts

### CODE REVIEW

File: packages/g6/src/item/item.ts [hunk 1]
```diff
@@ -112,8 +112,8 @@ export default abstract class Item implements IItem {
     this.stateMapper = stateMapper;
     this.displayModel = this.getDisplayModelAndChanges(model).model;
     this.renderExtensions = renderExtensions;
+    const type = this.type === 'node' ? 'circle-node' : 'line-edge';
```
Comment (Yanyan-Wang) [inline packages/g6/src/item/item.ts:115]: The type defined in this.displayModel.data has the first priority. 'cirlce-node' or 'line-edge' is the default value for 'node' or 'edge' respectively. This modification seems that the type will be fixed to 'circle-node' or 'line-edge', leading to the configured type in node/edge data does not take effect.

File: packages/g6/src/runtime/controller/item.ts [hunk 2]
```diff
@@ -317,9 +317,21 @@ export class ItemController {
           );
         }
         const node = itemMap[id] as Node;
-        const innerModel = graphCore.getNode(id);
+        // 'id' variable is always string in here, but one in user data is number, possibly.        
+        let innerModel
+        try {
+          innerModel = graphCore.getNode(id);
+        } catch (e) {
+          innerModel = graphCore.getNode(Number(id))
```
Comment (Yanyan-Wang) [inline packages/g6/src/runtime/controller/item.ts:325]: 👍

**Comment by Yanyan-Wang** [comment]
I think the TypeError of type 'circle' is caused by unregistered type name 'circle' in stdLib: https://github.com/antvis/G6/blob/dcb82900dba28f83532fa3ee39727970afdeaf67/packages/g6/src/stdlib/index.ts#L59

Since 'circle' is renamed to 'circle-node' in v5 to avoid the ambiguity of circle shape and circle type node.

If user want to custom a new node type, maybe named 'circle, 'xx-circle', or any string else, register it to the lib like: https://github.com/antvis/G6/blob/dcb82900dba28f83532fa3ee39727970afdeaf67/packages/g6/tests/unit/node-spec.ts#L312

If you want to add built-in type of node, maybe 'rect-node', register it to the stdLib as 'circle-node': https://github.com/antvis/G6/blob/dcb82900dba28f83532fa3ee39727970afdeaf67/packages/g6/src/stdlib/index.ts#L59


P.S. We only plan to add 'circle-node', 'sphere-node'(for 3d), 'rect-node' as built-in node types, they will be registered in G6's source code like the first link above. And other node types which are built-in in v4, e.g. 'star-node', 'triangle-node', etc. will also be implemented as ['circle-node' class here](https://github.com/antvis/G6/blob/v5/packages/g6/src/stdlib/item/node/circle.ts), but won't be registered to stdLib by default as the first link. When user want to use the 'star-node', they need to register it manually before using:

```
// importing the StarNode class from g6, which is not registered but has been defined in official code
import { StarNode } from '@antv/g6';

// register it manually before using
const ExtendedGraph = extend(G6.Graph, {
      nodes: {
        'star-node': StarNode,
      },
});

const graph = new ExtendedGraph({
  node: {
    // star-node can be configured only if it is registered before. 'circle-node' will take effect otherwise as default value
    type: 'star-node', 
  }
})
```


This is quiet different from v4, and this will bring tree-shaking effect.

File: packages/g6/src/runtime/controller/item.ts (hunk 2 above)
Comment (zqqcee) [reply packages/g6/src/runtime/controller/item.ts:325]: > 👍

So is this change acceptable？Should I create a new PR to solve this, independently?

**Comment by zqqcee** [comment]
> lib

Oops! Sorry, I misunderstood the meaning and caused the bug. But from your description, it's different from v4. 
I think you should note this change in the new documentation for v5. Thanks for your quick reply!

**Comment by Yanyan-Wang** [comment]
> > lib
> 
> Oops! Sorry, I misunderstood the meaning and caused the bug. But from your description, it's different from v4. I think you should note this change in the new documentation for v5. Thanks for your quick reply!

Yes, we will prepare a doc to diff v4 and v5 ASAP.

File: packages/g6/src/runtime/controller/item.ts (hunk 2 above)
Comment (Yanyan-Wang) [reply packages/g6/src/runtime/controller/item.ts:325]: > 

Maybe discard changes for the node type and keep this change in this pull request. Or a new PR is also ok. Just take your pick~

Comment (zqqcee) [reply packages/g6/src/runtime/controller/item.ts:325]: Hola, I've created a new PR to solve this problem

**Comment by Yanyan-Wang** [comment]
So this PR will be closed since the new one has solved all these problems https://github.com/antvis/G6/pull/4608

**Pull request closed without merging**